lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -299,6 +386,52 @@ bool IsDefaultActorRequest::evaluate(
   return true;
 }
 
+bool IsDistributedActorRequest::evaluate(
+    Evaluator &evaluator, NominalTypeDecl *nominal) const {
+  // Protocols are actors if their `Self` type conforms to `DistributedActor`.
+  if (auto protocol = dyn_cast<ProtocolDecl>(nominal)) {
+    // Simple case: we have the `DistributedActor` protocol itself.
+    if (protocol->isSpecificProtocol(KnownProtocolKind::DistributedActor))
+      return true;
+
+    auto actorProto = nominal->getASTContext().getProtocol(
+        KnownProtocolKind::DistributedActor);
+    if (!actorProto)
+      return false;
+
+    auto selfType = Type(protocol->getProtocolSelfType());
+    auto genericSig = protocol->getGenericSignature();
+    if (!genericSig)
+      return false;
+
+    return genericSig->requiresProtocol(selfType, actorProto);
+  }
+
+  // Class declarations are 'distributed actors' if they are declared with 'distributed actor'
+  if(!dyn_cast<ClassDecl>(nominal))
+    return false;
+
+  auto distributedAttr = nominal->getAttrs()
+      .getAttribute<DistributedActorAttr>();
+  return distributedAttr != nullptr;
+}
+
+bool IsDistributedFuncRequest::evaluate(
+    Evaluator &evaluator, FuncDecl *func) const {
+  // Check whether the attribute was explicitly specified.
+  if (auto attr = func->getAttrs().getAttribute<DistributedActorAttr>()) {
+//    // Check for well-formedness.
+//    if (checkDistributedFunc(func, /*diagnose=*/true)) {
```

Comment:
You could move the validation of `distributed` here, from the attribute checker. However, I do not think you should call `checkDistributedFunc` here, because that requires this otherwise lightweight check to fully resolve the types of its parameters and result type.

Reply:
Okey I'll follow up on this.